### PR TITLE
Introduce Futex for backpressure on Wards

### PR DESCRIPTION
**Pull request by shayanhabibi**

### CODE REVIEW

**Comment by shayanhabibi** [comment]
@disruptek the tward test seems to cork it; afaict this shouldnt be related to loony. Can you have a gander at it?

File: loony.nim [hunk 1]
```diff
@@ -283,7 +284,7 @@ proc popImpl[T](queue: LoonyQueue[T]; forcedCoherance: static bool = false): T =
     var (tail, curr) = tailAndMane queue
     if isEmptyImpl(curr, tail):
       # Queue was empty; nil can be caught in cps w/ "while cont.running"
-      return nil
+      return
```
Comment (disruptek) [inline loony.nim]: It's better style to explicitly `return nil` here.

File: loony.nim [hunk 2]
```diff
@@ -215,8 +215,9 @@ proc pushImpl[T](queue: LoonyQueue[T], el: T,
   # Ensure all writes in STOREBUFFER are committed. By far the most costly
   # primitive; it will be preferred while proving safety before working towards
   # optimisation by atomic reads/writes of cache lines related to el
-  when forcedCoherance:
-    atomicThreadFence(ATOMIC_RELEASE)
+  # when forcedCoherance:
+  #   atomicThreadFence(ATOMIC_RELEASE)
+  atomicThreadFence(ATOMIC_RELEASE)
```
Comment (disruptek) [inline loony.nim]: May as well remove it; if you leave it, it's useful as documentation: _we don't do this anymore because..._

File: loony.nim [hunk 3]
```diff
@@ -309,9 +310,15 @@ proc popImpl[T](queue: LoonyQueue[T]; forcedCoherance: static bool = false): T =
...
 
+          # when T is ref:
+          #   var res = cast[Loonatic[T]](prev and SLOTMASK)
+          #   result = extract res
+          # else:
           result = cast[T](prev and SLOTMASK)
-          assert not result.isNil
-          when result is ref:
+          when T is ref or T is pointer:
+            assert not result.isNil
```
Comment (disruptek) [inline loony.nim]: If you're going to potentially raise a defect in someone else's code, please add a string argument explaining to the user what the problem is.  I would prefer to just remove it and cover misbehavior with a unit test.

Is it likely someone passed a `T` that is not `ref` or `pointer`?

File: loony/ward.nim [hunk 4]
```diff
@@ -44,7 +49,7 @@ converter toWardFlags*(flags: set[WardFlag]): WardFlags =
   when nimvm:
     for flag in items(flags):
       block:
-        if flag == Pausable:
+        if flag == Pausable or flag == PoolWaiter:
```
Comment (disruptek) [inline loony/ward.nim]:
```suggestion
        if flag in {Pausable, PoolWaiter}:
```

File: loony/ward.nim [hunk 5]
```diff
@@ -59,7 +64,7 @@ converter toWardFlags*(flags: set[WardFlag]): WardFlags =
         PushPausable in flags and
         not (Pausable in flags):
       result = result or (1'u16 shl Pausable.ord)
-    if flags.contains Pausable:
+    if flags.contains(Pausable) or flags.contains(PoolWaiter):
```
Comment (disruptek) [inline loony/ward.nim]:
```suggestion
    if flags * {Pausable, PoolWaiter} != {}:
```

**Review by disruptek — APPROVED** [review]
I didn't test it, but made some minor style suggestions.

Also, I'd get in the habit of `import std/os` and `import pkg/balls` as opposed to unqualified identifiers.

File: loony.nim (hunk 3 above)
Comment (shayanhabibi) [reply loony.nim]: I originally had an opposing assertion which would assert that non-pointers would be compared to a default of their type.

**Comment by shayanhabibi** [comment]
i'll do destroy procs and a ward test suite that doesn't trigger arc undefined behaviour and then merge it

**Comment by shayanhabibi** [comment]
Welp still haven't done the ward test suite but I have added and tested the destroy proc. I think a lot of it is unnecessary since the destroy proc also goes through the effort of decref'in any refs still in there but might as well.

**Pull request merged**